### ISSUE TITLE
Java builder causes constant rebuilds if derived files are used

### ISSUE THREAD

**Issue by djh82** [SCons/scons#3725]
When using derived Java files, i.e. files generated by another builder (in our case protoc, but I suspect a simple case could be created with SubstFile?), the Java builder does not take into account the package they belong to and thus expects the resultant class files to appear in the root.  The classes actually appear in the package structure and thus on the next build SCons cannot find them and causes another rebuild.

Using python 3.6.8 and SCons 3.1.2 (scons-local) on Ubuntu 20.04.

I can try and contrive a simple example if necessary.

**Comment by bdbaddog** [comment]
Yes. We'd need a simple reproducer in order to create a test for any fix.

**Comment by djh82** [comment]
A simple SConstruct would be:

```python
env = Environment()
derived_file = env.Substfile('test/File.java.in', SUBSTDICT={})
env.Java('classes', derived_file)
```

Where test/File.java.in is:

```java
package test;

class File {}
```

`scons --debug=explain` outputs:

`scons: building 'classes/File.class' because it doesn't exist`

But the class has actually, correctly, been output to classes/test/File.class.
